site/app/libraries/DateUtils.php

```diff
@@ -82,21 +82,52 @@ public static function validateTimestamp($timestamp) {
     /**
      * Parses a date string into a \DateTime object, or does nothing if $date is already a \DateTime object
      *
-     * @param $date \DateTime|string The date to parse
-     * @param $time_zone \DateTimeZone The timezone to parse with
+     * @param \DateTime|string $date The date to parse
+     * @param \DateTimeZone $time_zone
      * @return \DateTime The parsed date
-     * @throws \InvalidArgumentException If $date is not a string or a \DateTime
-     * @throws \Exception If $date is a string, but not a valid date string
+     * @throws \InvalidArgumentException If $date is not a string or a \DateTime, or not a valid \DateTime string
      */
     public static function parseDateTime($date, \DateTimeZone $time_zone) {
         if ($date instanceof \DateTime) {
             return $date;
         } else if (gettype($date) === 'string') {
-            return new \DateTime($date, $time_zone);
+            try {
+                return new \DateTime($date, $time_zone);
+            } catch (\Exception $e) {
+                throw new \InvalidArgumentException('Invalid DateTime Format');
+            }
         } else {
             throw new \InvalidArgumentException('Passed object was not a DateTime object or a date string');
         }
     }
+
+    /**
+     * Parses a date string into a \DateTime object using regex.  This allows dates to be year >9999
+     * Note: This is designed so that dates larger than year 9999 can be loaded from the db without exception.
+     * Format YYYYY-MM-DD HH:mm:ssZ
+     *
+     * @param string $date_time
+     * @param \DateTimeZone|null $default_time_zone The default timezone to use if none provided
+     * @return DateTime
+     */
+    public static function parseDateTimeLong(string $date_time, $default_time_zone = null) {
+        $matches = [];
+        if (!preg_match('/^([0-9]{4,5})-([0-9]{2})-([0-9]{2}) ([0-9]{2}):([0-9]{2}):([0-9]{2})(\.[0-9]+)?([-|+]?[0-9]{2})?/', $date_time, $matches)) {
+            throw new \InvalidArgumentException('Invalid DateTime Format');
+        }
+
+        $ms = $matches[7] ?? '';
+        $timezone = $matches[8] ?? '';
+        if ($ms === '') {
+            $ms = 0;
+        }
+
+        $date = new \DateTime('now', ($timezone !== '') ? new DateTimeZone($timezone) : $default_time_zone);
+        $date->setDate($matches[1], $matches[2], $matches[3]);
+        $date->setTime($matches[4], $matches[5], $matches[6], $ms);
```

Comment:
This function throws a warning on PHP 7 as the millisecond parameter (fourth argument) was added in 7.1. Given that there's no real point to setting things at the microsecond level as we don't care about anything at that scale, no point worrying about it.